enhancements/cluster-logging/loki-gateway-configuration.md

```diff
@@ -0,0 +1,498 @@
+
+In this mode it is mandatory to provide a static list of per-tenant OIDC configuration as well as static RBAC. The LokiStack gateway deployment will be deployed with access to a static Rego file to match the RbacSpec. The static Rego file represents an in-process OPA endpoint.
+
+The LokiStack CR will look like this:
+
+```yaml
+tenants:
+    mode: static
+    authentication:
+    - name: tenant-a
+      oidc:
+        issuerURL: https://127.0.0.1:5556/dex
+        redirectURL: https://localhost:8443/oidc/tenant-a/callback
+        usernameClaim: test
+        groupClaim: test
+    authorization:
+      roleBindings:
+      - name: tenant-a
+        roles:
+        - read-write
+        subjects:
+        - kind: user
+          name: admin@example.com
+      roles:
+      - name: read-write
+        permissions:
+        - read
+        - write
+        resources:
+        - logs
+        tenants:
+        - tenant-a
+```
+
+The LokiStack Gateway deployment will need to create a static rego file which will include:
+
+```rego
+package lokistack
+
+import input
+import data.roles
+import data.roleBindings
+
+default allow = false
+
+allow {
+  some roleNames
+  roleNames = roleBindings[matched_role_binding[_]].roles
+  roles[i].name == roleNames[_]
+  roles[i].resources[_] = input.resource
+  roles[i].permissions[_] = input.permission
+  roles[i].tenants[_] = input.tenant
+}
+
+matched_role_binding[i] {
+  roleBindings[i].subjects[_] == {"name": input.subject, "kind": "user"}
+}
+
+matched_role_binding[i] {
+  roleBindings[i].subjects[_] == {"name": input.groups[_], "kind": "group"}
+}
+```
+
+### LokiStack CR configuration - _Dynamic Mode_
+
+In this mode it is mandatory to provide a static list of per-tenant OIDC configuration and a URL to an OPA endpoint. Latter is used to delegate the subject's information to access the requested tenant.
+
+_Note:_ Mixing static RBAC with dynamic mode results in a LokiStack CR degraded condition.
+
+The LokiStack CR will look like this:
+
+```yaml
+tenants:
+    mode: dynamic
+    authentication:
+    - name: tenant-a
+      oidc:
+        issuerURL: https://127.0.0.1:5556/dex
+        redirectURL: https://localhost:8443/oidc/tenant-a/callback
+        usernameClaim: test
+        groupClaim: test
+    authorization:
+      opa:
+        url: http://127.0.0.1:8080/v1/data/lokistack/allow
+```
+
+### LokiStack CR configuration - _Openshift Logging Mode_
+
+In this mode, user will need to do zero configuration as everything will be auto-configured by Loki operator itself.
+
+After the new [data model](https://github.com/openshift/enhancements/blob/master/enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md#data-model) was introduced, the logs from individual namespaces share common index. This was done because of multi-tenant feature of Openshift Logging which enables namespace-based tenancy since Openshift maintains this global view on resourses.
```

Comment:
This paragraph looks to be out of context. The new data model is Elasticsearch-only specific. The separation in three big buckets `application`, `infrastructure` and `audit` is more of a conceptual one that guides how to the collector and forwarder in OpenShift Logging categorizes logs. We continue using these three categories as three big tenants in Loki, because using namespaces as tenants will make our chunks too small. We ensure namespace-scoping by adding them as label-selectors, e.g.:

Let's say a user in OpenShift requires access to the logs of pod `my-pod` on namespace `my-project`, the call to lokistack will look like (translated in curl call for easier comprehension): 
```bash
$ curl -G -s  "http://lokistack.svc:8080/api/logs/v1/application/loki/api/v1/query_range" --data-urlencode 'query={pod_name="my-pod"}' --data-urlencode 'step=300' 
```

The lokistack-gateway will ask the authorizer sidecar `opa-openshift` and will retrieve the list of projects the user has access to and in turn amend the query from:
`{pod_name="my-pod"}`
to 
`{pod_name="my-pod", namespace=~"my-project|my-ther-project"}`

Thus the final upstream call to loki looks like (translated in curl call for easier comprehension):
```shell
$ curl -G -s -H "X-Scope-Org-ID: application"  "http://query-frontend.svc:8080/loki/api/v1/query_range" --data-urlencode 'query={pod_name="my-pod", namespace=~"my-project|my-ther-project"}' --data-urlencode 'step=300' 
```